### PR TITLE
Storage: Convert replica's and store's RWMutexes to Mutex

### LINKED ISSUE #3722: Mutex instead of RWMutex on Replica and Store
[from this discussion thread](https://github.com/cockroachdb/cockroach/pull/3710#issuecomment-170221620) via @bdarnell 

>
I would definitely make Replica use a Mutex. Store.mu makes a little more sense as a RWMutex since we intend to increase parallelism in the future, but it might be worth changing it to a Mutex for more immediate deadlock detection. Store.mu is on my hit list anyway; I need to refactor locking and the use of the processRaft goroutine.
>
@dt I've wanted to replace all our sync.Mutexes with a wrapper that adds an AssertHeld method; we could use a similar technique with build tags to replace RWMutex with a stricter version in some builds.
>
Maybe one day I'll get around to writing a go version of http://clang.llvm.org/docs/ThreadSafetyAnalysis.html

### PR DESCRIPTION
**Pull request by BramGruneir**
Fixes #3722 

commit 1) convert replica's sync.RWMutex to a sync.Mutex
commit 2) convert store's  sync.RWMutex to a sync.Mutex and embed it instead of using the mu variable
commit 3) turn replica's sync.Mutex into an anonymous struct
commit 4) turn store's sync.Mutex into an anonymous struct

I think I caught the comments about holding a read lock as well.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3747)

### CODE REVIEW

**Comment by BramGruneir** [comment]
I'm going to do a bit more work on this and convert both to embedded anonymous structs.

**Comment by tschottdorf** [comment]
LGTM, glad to see that went over so smoothly. Did you run the race tests for a bit to check for "obvious" deadlocks?

---

Reviewed 7 of 7 files at r1, 3 of 3 files at r2, 9 of 9 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[storage/replica_raftstorage.go, line 243 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3747#-K7mkjaSk-qt0QEqI3kp:-K7mkjaT9uyF5wNc4b1l:-1386797542) ([raw file](https://github.com/cockroachdb/cockroach/blob/543544f2e72246f401895d70f06226581dbbc7d4/storage/replica_raftstorage.go#L243)):*
the lock can only be acquired internally (since `mu` is unexported), which suggests that you should be able to rename `FirstIndex -> firstIndexLocked` and `GetFirstIndex -> FirstIndex`?

---

*[storage/replica_raftstorage.go, line 406 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3747#-K7mkuFcEK7X_bJSvTCY:-K7mkuFcEK7X_bJSvTCZ:294932679) ([raw file](https://github.com/cockroachdb/cockroach/blob/543544f2e72246f401895d70f06226581dbbc7d4/storage/replica_raftstorage.go#L406)):*
same as for `GetFirstIndex`?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3747#-:-K7ml06cbGBHICuS1vWc:-1339923904)*

**Comment by BramGruneir** [comment]
of course, working on one more commit for this PR
converting store.lock to a anon struct like in replica 

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[storage/replica_raftstorage.go, line 243 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3747#-K7mkjaSk-qt0QEqI3kp:-K7mld2jusBj5YqpUEb7:463158281) ([raw file](https://github.com/cockroachdb/cockroach/blob/543544f2e72246f401895d70f06226581dbbc7d4/storage/replica_raftstorage.go#L243)):*
sadly, no.  FirstIndex is part of the raft.Storage interface

---

*[storage/replica_raftstorage.go, line 406 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3747#-K7mkuFcEK7X_bJSvTCY:-K7mlp4JCBEUuEVmpWu8:678021867) ([raw file](https://github.com/cockroachdb/cockroach/blob/543544f2e72246f401895d70f06226581dbbc7d4/storage/replica_raftstorage.go#L406)):*
same

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3747#-:-K7mlSRKb_gJBlRg7NOp:1259105706)*

**Comment by BramGruneir** [comment]
Ok, I think this is all ready to go.  PTAL

**Comment by tschottdorf** [comment]
LGTM

---

Reviewed 4 of 4 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3747#-:-K7p1GFo_ooXz3k3ToBQ:676834233)*

File: storage/helpers_test.go [hunk 1]
```diff
@@ -62,30 +59,15 @@ func (s *Store) DisableRaftLogQueue(disabled bool) {
...
-	//
-	// End of (*Store).mu.Lock() danger zone
-	//
-	// (*Store).mu.RLock()
-	//
-	// See http://play.golang.org/p/xTJiAiRfYf
-	replicas := make([]*Replica, 0, len(s.replicas))
-	s.mu.RLock()
-	for _, r := range s.replicas {
+	// This copy allows calling s.Unlock() before calling MaybeAdd, which is
```
Comment (tamird) [inline storage/helpers_test.go]: this comment is no longer accurate AFAICT. there's no need for a concurrent call at all; this code will deadlock all by itself.

File: storage/replica_command.go [hunk 2]
```diff
@@ -1392,9 +1394,11 @@ func (r *Replica) splitTrigger(batch engine.Engine, split *roachpb.SplitTrigger)
 	// never make it to the new Range (see #3148), so all we grab here is
```
Comment (tamird) [inline storage/replica_command.go]: comment outdated

File: storage/replica_raftstorage.go [hunk 3]
```diff
@@ -240,8 +240,16 @@ func (r *Replica) FirstIndex() (uint64, error) {
 	return ts.Index + 1, nil
 }
 
+// GetFirstIndex is the same function as FirstIndex but it does not require
+// that the replica lock is held.
+func (r *Replica) GetFirstIndex() (uint64, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	return r.FirstIndex()
```
Comment (tamird) [inline storage/replica_raftstorage.go]: should `r.FirstIndex()` be unexported or renamed `FirstIndexLocked()`?

File: storage/replica_raftstorage.go [hunk 4]
```diff
@@ -394,6 +402,14 @@ func (r *Replica) Snapshot() (raftpb.Snapshot, error) {
 	}, nil
 }
 
+// GetSnapshot is the same function as Snapshot but it does not require the
+// replica lock to be held.
+func (r *Replica) GetSnapshot() (raftpb.Snapshot, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	return r.Snapshot()
```
Comment (tamird) [inline storage/replica_raftstorage.go]: same comment about exporting or renaming to *Locked

File: storage/replica_test.go [hunk 5]
```diff
@@ -665,7 +665,9 @@ func TestRangeTSCacheLowWaterOnLease(t *testing.T) {
 	tc.manualClock.Increment(int64(DefaultLeaderLeaseDuration + 1))
 	now := roachpb.Timestamp{WallTime: tc.manualClock.UnixNano()}
 
-	baseRTS, _ := tc.rng.tsCache.GetMax(roachpb.Key("a"), nil /* end */, nil /* txn */)
+	tc.rng.mu.Lock()
```
Comment (tamird) [inline storage/replica_test.go]: was the absence of this lock in old code a mistake?

File: storage/store.go [hunk 6]
```diff
@@ -380,29 +382,33 @@ func NewStore(ctx StoreContext, eng engine.Engine, nodeDesc *roachpb.NodeDescrip
...
-			ShouldEvict: func(size int, key, value interface{}) bool {
-				return size > maxReplicaDescCacheSize
-			},
-		}),
 		wakeRaftLoop:      make(chan struct{}, 1),
-		pendingRaftGroups: map[roachpb.RangeID]struct{}{},
 		raftRequestChan:   make(chan *RaftMessageRequest, raftReqBufferSize),
 	}
 
+	s.mu.Lock()
```
Comment (tamird) [inline storage/store.go]: isn't this excessive?

Comment (BramGruneir) [inline storage/store.go]: Possibly, but I don't see how it hurts.

File: storage/replica_raftstorage.go (hunk 3 above)
Comment (BramGruneir) [inline storage/replica_raftstorage.go]: Can't, sadly.  FirstIndex is part of the raft.Storage interface.

File: storage/replica_raftstorage.go (hunk 4 above)
Comment (BramGruneir) [inline storage/replica_raftstorage.go]: ditto

File: storage/replica_test.go (hunk 5 above)
Comment (BramGruneir) [inline storage/replica_test.go]: yes, seems so.  tsCache was meant to be protected, but we didn't actually do so.

File: storage/helpers_test.go (hunk 1 above)
Comment (BramGruneir) [inline storage/helpers_test.go]: Indeed, removed.

File: storage/replica_command.go (hunk 2 above)
Comment (BramGruneir) [inline storage/replica_command.go]: removed

**Comment by tamird** [comment]
Reviewed 2 of 7 files at r5.
Review status: 5 of 10 files reviewed at latest revision, 3 unresolved discussions.

---

*[storage/helpers_test.go, line 62 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3747#-K7lz0m8exO3cYzM6nvp-r4-62:-K7qdRN474igN-lGhTls:-1912812745) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ff5e93a1797d02bdcb852b91b078acab3ff7e70/storage/helpers_test.go#L62)):*
We still need a comment explaining why we make a copy, but it can be a simple discussion of re-entrance.

---

*[storage/replica_raftstorage.go, line 243 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3747#-K7mkjaSk-qt0QEqI3kp:-K7qdAp1APgzqKEWLvpl:-370394068) ([raw file](https://github.com/cockroachdb/cockroach/blob/543544f2e72246f401895d70f06226581dbbc7d4/storage/replica_raftstorage.go#L243)):*
so raft is calling this without acquiring the lock?

---

*[storage/replica_raftstorage.go, line 410 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3747#-K7lz0m8exO3cYzM6nvn-r4-410:-K7qdJka5jxyxyZoit5U:607662479) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ff5e93a1797d02bdcb852b91b078acab3ff7e70/storage/replica_raftstorage.go#L410)):*
so raft doesn't hold the lock while calling this?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3747)*

**Comment by BramGruneir** [comment]
Review status: 5 of 10 files reviewed at latest revision, 3 unresolved discussions.

---

*[storage/helpers_test.go, line 62 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3747#-K7lz0m8exO3cYzM6nvp-r4-62:-K7r-pxP5yqfq9iN_SwD:1989624981) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ff5e93a1797d02bdcb852b91b078acab3ff7e70/storage/helpers_test.go#L62)):*
I disagree.  It's a pretty standard pattern now.

---

*[storage/replica_raftstorage.go, line 243 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3747#-K7mkjaSk-qt0QEqI3kp:-K7r-thMK0-b7wCAS1Bw:376451485) ([raw file](https://github.com/cockroachdb/cockroach/blob/543544f2e72246f401895d70f06226581dbbc7d4/storage/replica_raftstorage.go#L243)):*
All calls to raft's rawnode are already locked. See the comment at the top of the file.

---

*[storage/replica_raftstorage.go, line 410 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3747#-K7lz0m8exO3cYzM6nvn-r4-410:-K7r0AmDL8pw4zIehGM4:-1152088624) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ff5e93a1797d02bdcb852b91b078acab3ff7e70/storage/replica_raftstorage.go#L410)):*
ditto, raft doesn't acquire the lock, it is already held.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3747)*

**Comment by tamird** [comment]
Review status: 5 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[storage/helpers_test.go, line 62 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3747#-K7lz0m8exO3cYzM6nvp-r4-62:-K7r0qX1s3AYA_35HEZu:-662248425) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ff5e93a1797d02bdcb852b91b078acab3ff7e70/storage/helpers_test.go#L62)):*
Really? Where else is it done? Here's a place that earned a comment https://github.com/cockroachdb/cockroach/blob/master/gossip/infostore.go#L214

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3747)*

**Comment by BramGruneir** [comment]
Review status: 4 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[storage/helpers_test.go, line 62 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3747#-K7lz0m8exO3cYzM6nvp-r4-62:-K7r1YWQzNIs5l6_2mnT:-1337438579) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ff5e93a1797d02bdcb852b91b078acab3ff7e70/storage/helpers_test.go#L62)):*
I'll add a comment in.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3747)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r6.
Review status: 5 of 10 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3747)*

**Pull request merged**